working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+parameters prefaced with the word `required` (e.g. `int Function(int, {int?
+y}, required {int z})`.
+
+The grammar of function/method/closure declarations is extended to allow an
+additional group of named parameters prefaced with the word `required`
+(e.g. `int f(int x, {int?  y}, required {int z})`.
+
+
+### Grammatical ambiguities and clarifications.
+
+#### Nested nullable types
+
+The grammar for types does not allow multiple successive `?` operators on a
+type.  That is, the grammar for types is
+
+```
+type' ::= functionType
+          | qualified typeArguments? 
+
+type ::= type' `?`?
+```
+
+#### Conditional expression ambiguities
+
+Conditional expressions inside of braces are ambiguous between sets and maps.
+That is, `{ a as int ? - 3 : 3 }` can be parsed as a set literal `{ (a as int)
+? - 3 : 3 }` or as a map literal `{ (a as int ?) - 3 : 3 }`.  Parsers will
+prefer the former parse over the latter.
+
+The same is true for `{ a is int ? - 3 : 3 }`.
+
+The same is true for `{ int ? - 3 : 3 }` if we allow this.
+
+
+## Static semantics
+
+
+### Static errors
+
+We say that a type `T` is **nullable** if `Null <: T`.  This is equivalent to
+the syntactic criterion that `T` is any of:
+  - `Null`
+  - `S?` for some `S`
+  - `FutureOr<S>` for some `S` where `S` is nullable
+  - `dynamic`
+  - `void`
+
+We say that a type `T` is **non-nullable** if `T <: Object`.  This is equivalent
+to the syntactic criterion that `T` is any of:
+  - `Object`, `int`, `bool`, `Never`, `Function`
+  - Any function type
+  - Any class type or generic class type
+  - `FutureOr<S>` where `S` is non-nullable
+  - `X extends S` where `S` is non-nullable
+  - `X & S` where `S` is non-nullable
+
+Note that there are types which are neither nullable nor non-nullable.  For
+example `X extends T` where `T` is nullable is neither nullable nor
+non-nullable.
+
+We say that a type `T` is **potentially nullable** if `T` is not non-nullable.
+Note that this is different from saying that `T` is nullable.  For example, a
+type variable `X extends Object?` is a type which is potentially nullable but
+not nullable.
+
+We say that a type `T` is **potentially non-nullable** if `T` is not nullable.
+Note that this is different from saying that `T` is non-nullable.  For example,
+a type variable `X extends Object?` is a type which is potentially non-nullable
+but not non-nullable.
+
+It is an error to call a method, setter, getter or operator on an expression
+whose type is potentially nullable and not `dynamic`, except for the methods,
+setters, getters, and operators on `Object`.
+
+It is an error to read a field or tear off a method from an expression whose
+type is potentially nullable and not `dynamic`, except for the methods and
+fields on `Object`.
+
+It is an error to call an expression whose type is potentially nullable and not
+`dynamic`.
+
+It is an error for an instance field with no static initializer to have a
```

Comment:
What does "static initializer" mean?
Should perhaps be:

It is a compile-time error if a class has an instance field with a potentially nullable type and no initializer expression, and it has a non-redirecting generative constructor that does not intitialize that variable through an initializing formal or initializer list entry.

Reply:
done.